File: Source/Playwright.ts
import { GuildMember, Snowflake } from 'discord.js';
import DOMPurify, { Config } from 'isomorphic-dompurify';
import { type Browser, BrowserContext, chromium, type Page } from 'playwright';

import { Nullable } from '../Common/types';

import { Context } from './Context';

const PAGE_OPERATION_TIMEOUT = 5000;
const VALID_URL_PATTERN =
    /^https:\/\/(?:cdn\.discordapp\.com|media\.discordapp\.net|i\.imgur\.com)\//;

export type TPlaywright = Playwright;

interface RenderData {
    attachments?: string | string[];
    avatar: string;
    content: string;
    replyAvatar?: string;
    replyContent?: string;
    replyUsername?: string;
    roleIcon?: string;
    timestamp: string;
    username: string;
}

interface RenderOptions {
    data: RenderData;
    deviceScaleFactor?: number;
    html: string;
    selector?: string;
    viewport?: {
        height: number;
        width: number;
    };
}

interface TemplateCache {
    browser: Nullable<Browser>;
    lastCleanup: number;
    maxPages: number;
    operationQueue: Promise<void>;
    pages: Array<{
        context: BrowserContext;
        lastUsed: number;
        page: Page;
    }>;
}

export default class Playwright {
    private static instance: Playwright;
    private sanitizerConfig: Config;
    private templateCache: TemplateCache;

    private constructor() {
        this.templateCache = {
            browser: null,
            lastCleanup: Date.now(),
            maxPages: 3,
            operationQueue: Promise.resolve(),
            pages: [],
        };
        this.sanitizerConfig = {
            ALLOW_ARIA_ATTR: false,
            ALLOW_DATA_ATTR: false,
            ALLOW_UNKNOWN_PROTOCOLS: false,
            ALLOWED_ATTR: ['class', 'src', 'alt', 'title'],
            ALLOWED_TAGS: ['div', 'span', 'img'],
            FORBID_TAGS: ['script', 'style', 'iframe', 'object', 'embed', 'svg', 'math'],
            SAFE_FOR_TEMPLATES: true,
        };
    }

    public static getInstance(): Playwright {
        if (!Playwright.instance) {
            Playwright.instance = new Playwright();
        }
        return Playwright.instance;
    }

    public async cleanupCache(): Promise<void> {
        try {
            await this.cleanupOldPages();
            if (this.templateCache.browser) {
                await this.templateCache.browser.close();
                this.templateCache.browser = null;
            }
        } catch (error) {
            console.error('Error during cleanup:', error);
        }
    }

    public command(attribute: string): string {
        return `jsp-${attribute}`;
    }

    public async generateImage(options: RenderOptions): Promise<Buffer> {
        let context: BrowserContext | null = null;
        let page: null | Page = null;

        try {
            await this.ensureBrowserHealth();
            if (!this.templateCache.browser) {
                throw new Error('Browser not initialized');
            }

            context = await this.templateCache.browser.newContext({
                bypassCSP: true,
                deviceScaleFactor: options.deviceScaleFactor || 4,
                viewport: options.viewport || { height: 150, width: 480 },
            });

            page = await context.newPage();
            page.on('console', (msg) => console.log('Page log:', msg.text()));
            page.on('pageerror', (err) => console.error('Page error:', err));

            await page.addInitScript(() => {
                document.documentElement.style.backgroundColor = 'transparent';
                document.body.style.backgroundColor = 'transparent';
            });

            await page.setContent(options.html, {
                timeout: PAGE_OPERATION_TIMEOUT,
                waitUntil: 'networkidle',
            });

            await page.evaluate(() => document.fonts?.ready);
            const element = await page.$(options.selector || 'body');
            if (!element) throw new Error('Could not find target element');

            const boundingBox = await element.boundingBox();
            if (!boundingBox) throw new Error('Could not get element dimensions');

            await page.setViewportSize({
                height: Math.ceil(boundingBox.height),
                width: Math.max(520, Math.ceil(boundingBox.width)),
            });

            await page.evaluate(
                ({ data }: { data: RenderData }) => {
                    function framework(dom: Document, key: string, value): void {
                        const conditionalElements = dom.querySelectorAll(
                            `[data-exists-then-display="${key}"]`,
                        );
                        conditionalElements.forEach((element: HTMLElement) => {
                            const toggleClass = element.getAttribute('data-toggle-class');
                            if (!toggleClass) return;

                            const isEmpty =
                                value === undefined ||
                                value === null ||
                                (typeof value === 'string' && value.trim() === '');

                            if (isEmpty) {
                                element.classList.add(toggleClass);
                            } else {
                                element.classList.remove(toggleClass);
                            }
                        });

                        const elements = dom.querySelectorAll(`[data-bind="${key}"]`);
                        elements.forEach((element) => {
                            if (element instanceof HTMLImageElement) {
                                element.src = String(value);
                            } else if (element instanceof HTMLElement) {
                                if (
                                    typeof value === 'string' &&
                                    value.startsWith('<') &&
                                    value.endsWith('>')
                                ) {
                                    element.innerHTML = value;
                                } else {
                                    element.textContent = String(value);
                                }
                            }
                        });
                    }

                    return new Promise<void>((resolve) => {
                        Object.entries(data).forEach(([key, value]) => {
                            framework(document, key, value);
                        });

                        Promise.all(
                            Array.from(document.images).map((img) => {
                                if (img.complete) return Promise.resolve();
                                return new Promise((resolve) => {
                                    img.onload = resolve;
                                    img.onerror = resolve;
                                });
                            }),
                        ).then(() => resolve());
                    });
                },
                { data: options.data },
            );

            await page.evaluate(() => document.fonts?.ready);

            return (await element.screenshot({
                animations: 'disabled',
                omitBackground: false,
                scale: 'device',
                type: 'png',
            })) as Buffer;
        } catch (error) {
            console.error('Error generating image:', error);
            throw new Error(`Image generation failed: ${error.message}`);
        } finally {
            try {
                if (page) await page.close();
                if (context) await context.close();
            } catch (error) {
                console.error('Error during cleanup:', error);
            }
        }
    }

    public isValidUrl(url: string): boolean {
        return VALID_URL_PATTERN.test(url);
    }

    public async parseMentions(
        ctx: Context,
        guildId: Snowflake,
        text: string[],
    ): Promise<string[]> {
        for (let i = 0; i < text.length; i++) {
            if (text[i].startsWith('<@')) {
                const userId = text[i].slice(2, -1);
                try {
                    const guild = await ctx.guilds.fetch(guildId);
                    const member = (await guild.members.fetch(userId)) as GuildMember;
                    if (member) {
                        const name = member.displayName || member.user.username;
                        text[i] = `<span class="mention" data-user-id="${name}">@${name}</span>`;
                    }
                } catch (error) {
                    console.error('Error fetching member:', error);
                    continue;
                }
            }
        }
        return text;
    }

    public sanitize(text: string): string {
        return DOMPurify.sanitize(text, this.sanitizerConfig);
    }

    private async cleanupOldPages(): Promise<void> {
        const now = Date.now();
        this.templateCache.lastCleanup = now;
        try {
            await Promise.all(this.templateCache.pages.map((p) => p.context.close()));
            this.templateCache.pages = [];
        } catch (error) {
            console.error('Error during page cleanup:', error);
        }
    }

    private async ensureBrowserHealth(): Promise<void> {
        if (!this.templateCache.browser) {
            await this.initializeCache();
            return;
        }
        try {
            const testContext = await this.templateCache.browser.newContext();
            await testContext.close();
        } catch (error) {
            console.error('Browser health check failed:', error);
            await this.initializeCache();
        }
    }

    private async initializeCache(retryCount = 0): Promise<void> {
        try {
            if (this.templateCache.browser) {
                await this.cleanupCache();
            }
            const browser = await chromium.launch({
                args: [
                    '--disable-gpu',
                    '--disable-dev-shm-usage',
                    '--js-flags=--noexpose_wasm',
                    '--disable-background-networking',
                    '--disable-default-apps',
                    '--disable-extensions',
                    '--disable-sync',
                    '--disable-translate',
                    '--hide-scrollbars',
                    '--metrics-recording-only',
                    '--mute-audio',
                    '--no-first-run',
                    '--safebrowsing-disable-auto-update',
                ],
            });
            this.templateCache.browser = browser;
        } catch (error) {
            console.error(`Failed to initialize browser (attempt ${retryCount + 1}):`, error);
            if (retryCount < 2) {
                await new Promise((resolve) => setTimeout(resolve, 1000 * (retryCount + 1)));
                return this.initializeCache(retryCount + 1);
            }
            throw error;
        }
    }
}
export const playwright = Playwright.getInstance();

File: apps/webserver/src/main/kotlin/tara/tb1/pw/Playwright/Playwright.kt
package tara.tb1.pw.Playwright

import com.fasterxml.jackson.module.kotlin.convertValue
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.microsoft.playwright.*
import com.microsoft.playwright.options.*
import java.util.regex.Pattern
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.jsoup.Jsoup
import org.jsoup.safety.Safelist

private const val PAGE_OPERATION_TIMEOUT = 5000
private val VALID_URL_PATTERN =
        Pattern.compile(
                "^https://(?:cdn\\.discordapp\\.com|media\\.discordapp\\.net|i\\.imgur\\.com)/"
        )

data class RenderData(
        val attachments: List<String>? = null,
        val avatar: String,
        val content: String,
        val replyAvatar: String? = null,
        val replyContent: String? = null,
        val replyUsername: String? = null,
        val roleIcon: String? = null,
        val timestamp: String,
        val username: String,
        val usernameColor: String? = null,
        val replyUsernameColor: String? = null,
        val customData: Map<String, Any?> = emptyMap(),
        val userId: String? = null,
        val roleId: String? = null,
        val channelId: String? = null,
        val roleName: String? = null,
        val channelName: String? = null
)

data class RenderOptions(
        val data: RenderData,
        val deviceScaleFactor: Int = 4,
        val html: String,
        val selector: String? = null,
        val viewport: Viewport = Viewport(150, 480)
)

data class Viewport(val height: Int, val width: Int)

data class TemplateCache(
        var browser: Browser? = null,
        var lastCleanup: Long = System.currentTimeMillis(),
        val maxPages: Int = 3,
        val pages: MutableList<PageInfo> = mutableListOf()
)

data class PageInfo(
        val context: BrowserContext,
        val page: Page,
        var lastUsed: Long = System.currentTimeMillis()
)

object Playwright {
    private val templateCache = TemplateCache()
    private val safeList =
            Safelist.basic()
                    .addAttributes("div", "class", "style")
                    .addAttributes("span", "class", "style")
                    .addAttributes("img", "src", "alt", "title")

    init {
        safeList.addTags("div", "span", "img")
        safeList.removeTags("script", "style", "iframe", "object", "embed", "svg", "math")
    }

    suspend fun cleanupCache() =
            withContext(Dispatchers.IO) {
                try {
                    cleanupOldPages()
                    templateCache.browser?.close()
                    templateCache.browser = null
                } catch (error: Exception) {
                    System.err.println("Error during cleanup: ${error.message}")
                }
            }

    fun command(attribute: String): String = "jsp-$attribute"

    suspend fun generateImage(options: RenderOptions): ByteArray =
            withContext(Dispatchers.IO) {
                var context: BrowserContext? = null
                var page: Page? = null

                try {
                    ensureBrowserHealth()
                    requireNotNull(templateCache.browser) { "Browser not initialized" }

                    val contextOptions =
                            Browser.NewContextOptions()
                                    .setViewportSize(
                                            options.viewport.width,
                                            options.viewport.height
                                    )
                                    .setDeviceScaleFactor(options.deviceScaleFactor.toDouble())

                    context = templateCache.browser?.newContext(contextOptions)
                    page = context?.newPage()

                    page?.onConsoleMessage { msg -> println("Page log: ${msg.text()}") }
                    page?.onPageError { err -> System.err.println(err) }

                    page?.addInitScript(
                            """
            document.documentElement.style.backgroundColor = 'transparent';
            document.body.style.backgroundColor = 'transparent';
        """.trimIndent()
                    )

                    val setContentOptions =
                            Page.SetContentOptions()
                                    .setTimeout(PAGE_OPERATION_TIMEOUT.toDouble())
                                    .setWaitUntil(WaitUntilState.NETWORKIDLE)
                    page?.setContent(options.html, setContentOptions)

                    page?.evaluate("() => document.fonts?.ready")

                    if (!options.data.attachments.isNullOrEmpty()) {
                        page?.evaluate(
                                """
                            async () => {
                                const images = document.getElementsByTagName('img');
                                await Promise.all(
                                    Array.from(images).map(img =>
                                        img.complete ? 
                                            Promise.resolve() : 
                                            new Promise((resolve, reject) => {
                                                img.onload = resolve;
                                                img.onerror = reject;
                                            })
                                    )
                                );
                            }
                        """.trimIndent()
                        )
                    }

                    val element =
                            page?.querySelector(options.selector ?: "body")
                                    ?: throw IllegalStateException("Could not find target element")

                    val boundingBox =
                            element.boundingBox()
                                    ?: throw IllegalStateException(
                                            "Could not get element dimensions"
                                    )

                    page.setViewportSize(
                            maxOf(520, boundingBox.width.toInt()),
                            boundingBox.height.toInt()
                    )

                    val bindDataFunction =
                            """
            async function(data) {
                const processElements = (entryKey, entryValue) => {
                    if (entryKey === 'attachments' && Array.isArray(entryValue)) {
                        const attachmentsDiv = document.querySelector('[data-bind-attachments]');
                        if (attachmentsDiv && entryValue) {
                            attachmentsDiv.innerHTML = entryValue.join('\n');
                            attachmentsDiv.style.display = 'block';
                        }
                        return;
                    }
                    document.querySelectorAll(`[data-exists-then-display="${'$'}{entryKey}"]`).forEach(element => {
                        const toggleClass = element.getAttribute('data-toggle-class')
                        if (!toggleClass) return
                        
                        const isEmpty = !entryValue || (typeof entryValue === 'string' && !entryValue.trim())
                        element.classList.toggle(toggleClass, isEmpty)
                    })
                    document.querySelectorAll(`[data-bind-class="${'$'}{entryKey}"]`).forEach(element => {
                        if (element instanceof HTMLElement) {
                            element.className += ` ${'$'}{entryValue}`;
                        }
                    });

                    document.querySelectorAll(`[data-bind="${'$'}{entryKey}"]`).forEach(element => {
                        const value = String(entryValue || '')
                        if (element instanceof HTMLImageElement) {
                            element.src = value
                        } else if (element instanceof HTMLElement) {
                            element.innerHTML = value.startsWith('<') && value.endsWith('>') ? value : value
                        }
                    })
                    
                    document.querySelectorAll(`[data-bind-style="${'$'}{entryKey}"]`).forEach(element => {
                        if (element instanceof HTMLElement) {
                            const styleValue = String(entryValue || '')
                            element.style.cssText += styleValue
                        }
                    })

                    if (entryKey === 'usernameColor' || entryKey === 'replyUsernameColor') {
                        document.querySelectorAll(`[data-color="${'$'}{entryKey}"]`).forEach(element => {
                            if (element instanceof HTMLElement) {
                                element.style.color = String(entryValue || '#FFFFFF')
                            }
                        })
                    }
                }

                for (let [entryKey, entryValue] of Object.entries(data)) {
                    if (entryKey === 'customData') continue;
                    processElements(entryKey, entryValue)
                }

                if (data.customData) {
                    for (let [entryKey, entryValue] of Object.entries(data.customData)) {
                        processElements(entryKey, entryValue)
                    }
                }

                await Promise.all(Array.from(document.images).map(img =>
                    img.complete ? Promise.resolve() : new Promise(r => { img.onload = r; img.onerror = r; })
                ))
            }
        """.trimIndent()

                    val processedData =
                            options.data.copy(
                                    content =
                                            if (options.data.userId != null) {
                                                parseDiscordMentions(
                                                        options.data.content,
                                                        options.data.userId,
                                                        options.data.roleId ?: "",
                                                        options.data.channelId ?: "",
                                                        options.data.username,
                                                        options.data.roleName ?: "Role",
                                                        options.data.channelName ?: "channel",
                                                        options.data.customData
                                                )
                                            } else options.data.content,
                                    replyContent =
                                            if (options.data.userId != null &&
                                                            options.data.replyContent != null
                                            ) {
                                                parseDiscordMentions(
                                                        options.data.replyContent,
                                                        options.data.userId,
                                                        options.data.roleId ?: "",
                                                        options.data.channelId ?: "",
                                                        options.data.username,
                                                        options.data.roleName ?: "Role",
                                                        options.data.channelName ?: "channel",
                                                        options.data.customData
                                                )
                                            } else options.data.replyContent
                            )

                    val objectMapper = jacksonObjectMapper()
                    val dataMap: Map<String, Any?> = objectMapper.convertValue(processedData)
                    page.evaluate(bindDataFunction, dataMap)

                    page.evaluate("() => document.fonts?.ready")

                    val screenshotOptions =
                            ElementHandle.ScreenshotOptions()
                                    .setOmitBackground(false)
                                    .setType(ScreenshotType.PNG)

                    return@withContext element.screenshot(screenshotOptions)
                } catch (error: Exception) {
                    throw IllegalStateException("Image generation failed: ${error.message}")
                } finally {
                    try {
                        page?.close()
                        context?.close()
                    } catch (error: Exception) {
                        System.err.println("Error during cleanup: ${error.message}")
                    }
                }
            }

    fun isValidUrl(url: String): Boolean = VALID_URL_PATTERN.matcher(url).matches()

    fun sanitize(text: String): String = Jsoup.clean(text, "", safeList)

    private suspend fun cleanupOldPages() =
            withContext(Dispatchers.IO) {
                val now = System.currentTimeMillis()
                templateCache.lastCleanup = now
                try {
                    templateCache.pages.forEach { it.context.close() }
                    templateCache.pages.clear()
                } catch (error: Exception) {
                    System.err.println("Error during page cleanup: ${error.message}")
                }
            }

    private suspend fun ensureBrowserHealth() =
            withContext(Dispatchers.IO) {
                if (templateCache.browser == null) {
                    initializeBrowser()
                    return@withContext
                }
                try {
                    templateCache.browser?.newContext()?.use { testContext -> testContext.close() }
                } catch (error: Exception) {
                    System.err.println("Browser health check failed: ${error.message}")
                    initializeBrowser()
                }
            }

    private suspend fun initializeBrowser(retryCount: Int = 0): Browser =
            withContext(Dispatchers.IO) {
                try {
                    templateCache.browser?.close()

                    val playwright = com.microsoft.playwright.Playwright.create()
                    val launchOptions =
                            BrowserType.LaunchOptions()
                                    .setArgs(
                                            listOf(
                                                    "--disable-gpu",
                                                    "--disable-dev-shm-usage",
                                                    "--js-flags=--noexpose_wasm",
                                                    "--disable-background-networking",
                                                    "--disable-default-apps",
                                                    "--disable-extensions",
                                                    "--disable-sync",
                                                    "--disable-translate",
                                                    "--hide-scrollbars",
                                                    "--metrics-recording-only",
                                                    "--mute-audio",
                                                    "--no-first-run",
                                                    "--safebrowsing-disable-auto-update"
                                            )
                                    )

                    val browser = playwright.chromium().launch(launchOptions)
                    templateCache.browser = browser
                    browser
                } catch (error: Exception) {
                    System.err.println(
                            "Failed to initialize browser (attempt ${retryCount + 1}): ${error.message}"
                    )
                    if (retryCount < 2) {
                        kotlinx.coroutines.delay((retryCount + 1) * 1000L)
                        initializeBrowser(retryCount + 1)
                    } else {
                        throw error
                    }
                }
            }

    fun parseDiscordMentions(
            text: String,
            userId: String,
            roleId: String,
            channelId: String,
            userName: String = "User",
            roleName: String = "Role",
            channelName: String = "channel",
            customData: Map<String, Any?> = emptyMap()
    ): String {
        val pattern = Regex("<(@!?|@&|#)(\\d+)>")

        val objectMapper = jacksonObjectMapper()
        val mentionedUsers =
                customData["mentionedUsers"]?.let {
                    objectMapper.convertValue<Map<String, String>>(it)
                }
                        ?: emptyMap()
        val mentionedRoles =
                customData["mentionedRoles"]?.let {
                    objectMapper.convertValue<Map<String, String>>(it)
                }
                        ?: emptyMap()
        val mentionedChannels =
                customData["mentionedChannels"]?.let {
                    objectMapper.convertValue<Map<String, String>>(it)
                }
                        ?: emptyMap()

        return pattern.replace(text) { matchResult ->
            val type = matchResult.groupValues[1]
            val id = matchResult.groupValues[2]

            when (type) {
                "@!", "@" -> {
                    val mentionedName = if (id == userId) userName else mentionedUsers[id] ?: "User"
                    """<span class="mention user" data-id="$id">@${mentionedName}</span>"""
                }
                "@&" -> {
                    val mentionedRoleName =
                            if (id == roleId) roleName else mentionedRoles[id] ?: "Role"
                    """<span class="mention role" data-id="$id">@${mentionedRoleName}</span>"""
                }
                "#" -> {
                    val mentionedChannelName =
                            if (id == channelId) channelName else mentionedChannels[id] ?: "channel"
                    """<span class="mention channel" data-id="$id">#${mentionedChannelName}</span>"""
                }
                else -> matchResult.value
            }
        }
    }
}
